classlib/src/main/java/org/teavm/classlib/java/util/TAbstractMap.java

```diff
@@ -237,14 +237,18 @@ public boolean equals(Object obj) {
         if (size() != other.size()) {
             return false;
         }
-        for (TIterator<? extends TMap.Entry<K, V>> iter = entrySet().iterator(); iter.hasNext();) {
-            TMap.Entry<K, V> entry = iter.next();
-            if (!other.containsKey(entry.getKey())) {
-                return false;
-            }
-            if (!TObjects.equals(entry.getValue(), other.get(entry.getKey()))) {
-                return false;
+        try {
+            for (var it = entrySet().iterator(); it.hasNext();) {
+                TMap.Entry<K, V> entry = it.next();
+                if (!other.containsKey(entry.getKey())) {
+                    return false;
+                }
+                if (!TObjects.equals(entry.getValue(), other.get(entry.getKey()))) {
+                    return false;
+                }
             }
+        } catch (ClassCastException | NullPointerException e) {
```

Comment:
FYI: in JS BE these exceptions are thrown by VM *only* in strict mode, which is turned off by default, due to performance reasons. It's only turned on for tests. Usually it's a bad practice to rely on CCE and NPE as a normal contract, and *most* of the code I ever dealt with, follows this rule.

Reply:
I can explain to you this change. In contract for TreeMap and some other maps (ConcurrentHashMap for instance) it's possible to throw NPE or CCE during containsKey or get. equals method should never throw exceptions.
So, assuming the code
```
TreeMap<Integer, Integer> tm = new TreeMap();
tm.add(1, 1);
Map<String, String> m = Map.of("a", "b");
System.out.println(m.equals(tm));
```
in old code exception will be thrown instead of false. So either this fix or calls to other.containsKey() and other.get() should be guarded (via same try-catch) somehow from exceptions.

Reply:
Hmm, now get your point. In this case I assume that corresponding methods will return "false" and null. I'm assuming that in this case correct "false" will be returned to equals.

Reply:
Corresponding methods will produce UB (although there's not such thing in Java) like throwing native JS errors or returning values of unexpected types.